Topics/System_Design/Twitter_System_Design.md

```diff
@@ -0,0 +1,63 @@
+# Twitter Case Study: How to Design the Home Timeline for Massive Scale
+
+## Introduction
+Twitter's home timeline is in theory a very simple feature. Users are in a many-to-many relationship with other users. Users can post Tweets. Each follower of the user should see every Tweet posted by the people they follow in their home feed, sorted in chronological order.
+
+The challenge for Twitter is building a highly-available and responsive home timeline given their massive scale. As of November 2012, Twitter had:
+- 150M world wide active users
+- Over 300K Queries-per-Second (QPS) for timelines
+
+In conjunction with this scale, Raffi Krikorian notes in his [QCon Talk](https://www.infoq.com/presentations/Twitter-Timeline-Scalability/) that when a user tweets, their followers should be updated with the new tweet in 5 seconds or under. This article will explain why this is a non-trivial task and give a high-level overview of how Twitter designed their systems to accomodate their scale (as of 2012).
+
+## Approach 1: Naive SQL Query
+Below is a simplified SQL schema of Twitter's data model.
+
+**tweets**
+| id   |      author_id      | content  |
+|----------|:-------------:|------:|
+| 38927 | 1  | "Did you ever hear the tragedy of Darth Plagueis The Wise? I thought not." |
+| 38928 |    2   |   "I hate sand." |
+| 38929 | 3 |    "You were my brother Anakin!" |
+
+**users**
+| id   |      username      |  bio |
+|----------|:-------------:|------:|
+| 1 |  darth-palpatine | Unlimited Power! |
+| 2 |    anakin-skywalker   |   The Jedi are Evil |
+| 3 | obi-wan |  Hello there   |
+
+**follows**
+| follower_id   |      follows_id      |
+|----------|:-------------:|
+| 2 |  1 |
+| 2 |    3   |
+| 1 | 2 |
+
+Twitter's initial naive approach was to join the Tweet, User, and Follows tables in SQL upon a GET request from a client like so:
+```
+SELECT * FROM tweets
+JOIN users ON tweets.author_id = users.id
+JOIN follows ON follows_id = users.id
+WHERE follower_id = current_user.id;
+```
+
+You can quickly see how this approach became quite slow at over 300K QPS.
```

Comment:
Elaborate on this maybe?

Reply:
Do you think it's unclear? I felt like it was evident that JOINing the entire userbase with all Tweets and all follow relationship for every query would not be a good idea. But maybe that isn't clear, should I add that explanation?

Reply:
I guess it depends on who's reading. "You can quickly see how [JOINing the entire userbase with all Tweets ...] became quite slow at over 300K QPS"? Probably not a major issue since it is clear why if you have basic database knowledge